### PR TITLE
Issue/4274 auto full export

### PR DESCRIPTION
**Pull request by sanderr**
# Description

Still in draft

# Questions
Did I pick an appropriate cron library? If anyone prefers another one for some reason it should be relatively simple at this point to swap it. The ones I considered are the following:
- [crontab](https://pypi.org/project/crontab/): 286 stars, simple interface to get the number of seconds till next time point (exactly what we need), well documented which subset is supported, latest commit 1+ year old
- [croniter](https://pypi.org/project/croniter/): 83 stars, objects are generators (a bit more advanced than what we use it for), documentation OK, actively maintained (latest release last month), seems to use SemVer
- [python-crontab](https://pypi.org/project/python-crontab/): 73 stars, cool icon, doesn't seem to support parsing just the time-part of a cron expression

I went for `croniter` mostly because it seems sufficiently stable and is actively maintained for sure. I'm not sure if we should worry about the latter though, perhaps `crontab` just didn't receive updates because it is so stable?

closes *Add ticket reference here*

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [ ] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

**Comment by FloLey** [comment]
The PR is not linked to the ticket but to itself in the description :p

**Comment by sanderr** [comment]
Oops, thanks, fixed.

File: src/inmanta/util.py [hunk 1]
```diff
@@ -112,31 +115,130 @@ def handler(future):
...
+
+    @abstractmethod
+    def log(self, action: TaskMethod) -> None:
+        """
+        Log a message about the action being scheduled according to this schedule.
+        """
+        ...
+
+
+ScheduledTask = Tuple[TaskMethod, TaskSchedule]
```
Comment (arnaudsjs) [inline src/inmanta/util.py:151]: Not really change request, just a proposal: I think it would be an improvement if we turn this into an object. Now every client of the scheduler has pack/unpack this tuple in order to interact with the scheduler.

File: tests/server/test_compilerservice.py [hunk 2]
```diff
@@ -598,10 +599,47 @@ async def test_server_recompile(server, client, environment, monkeypatch):
...
+    assert versions["count"] == 5
+
+    # delete schedule, verify it is cancelled
+    await schedule_soon()
+    await client.environment_setting_delete(environment, id="auto_full_compile")
+    with pytest.raises(AssertionError, match="Bounded wait failed"):
+        await retry_limited(is_compiling, 4)
+    assert (await client.list_versions(environment)).result["count"] == 5
+
+    # override with schedule in far future, check that it doesn't trigger an immediate recompile
```
Comment (arnaudsjs) [inline tests/server/test_compilerservice.py:635]:
```suggestion
    # override with schedule in the past, check that it doesn't trigger an immediate recompile
```

File: tests/server/test_compilerservice.py [hunk 3]
```diff
@@ -598,10 +599,47 @@ async def test_server_recompile(server, client, environment, monkeypatch):
...
+    recent: datetime.datetime = datetime.datetime.now(datetime.timezone.utc) - datetime.timedelta(seconds=2)
+    cron_recent: str = "%d %d * * * %d" % (recent.minute, recent.hour, recent.second)
```
Comment (arnaudsjs) [inline tests/server/test_compilerservice.py:637]: This cron expression isn't correct, is it? Or am I missing something?

File: tests/test_util.py [hunk 4]
```diff
@@ -20,38 +20,184 @@
...
     length = len(i)
     await asyncio.sleep(0.1)
     assert len(i) == length
-    assert not sched._executing_tasks[action]
+    assert not scheduler._executing_tasks[action]
+    assert not (action, schedule) in scheduler._scheduled
     no_error_in_logs(caplog)
 
 
+async def test_scheduler_remove_same_action() -> None:
```
Comment (arnaudsjs) [inline tests/test_util.py:126]: This test case can use a docstring.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/server/protocol.py [hunk 5]
```diff
@@ -296,7 +296,29 @@ def get_handlers(self) -> List[routing.Rule]:
...
+
+        :param interval: The interval between executions of the task.
+        :param initial_delay: The delay to execute the task for the first time. If not set, interval is used.
+        """
+        self._sched.add_action(call, IntervalSchedule(float(interval), initial_delay))
+
+    def schedule_cron(self, call: TaskMethod, cron: str) -> None:
+        """
+        Schedule a task according to a cron specifier. Tasks with the same call and the same schedule are considered the same.
+        Clients that wish to be able to delete tasks should make sure to use a unique action object.
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/server/protocol.py:311]: I don't expect this to ever really be the case, but there was some undefined behavior here if the same action was scheduled multiple times. I've made it slightly more robust by including the schedule in the comparison, but mostly I've made an attempt to clarify the current behavior in the docstring.

Suppose two clients schedule the same task using the same schedule (using cron for simplicity though an interval schedule is more plausible):
```
def mytask() -> None:
    print("doing things")

def client_one() -> None:
    schedule_cron(mytask, IntervalSchedule(3600, initial_delay=0))

def client_two() -> None:
    schedule_cron(mytask, IntervalSchedule(3600, initial_delay=0))

client_one()
time.sleep(1800)
client_two()
```

The scheduler uses `(task, schedule)` as a key, so if one of the clients attempts to then delete the scheduled action, one of them will be deleted but there's no telling which.

Should I clarify this better? Or try to improve the handling of this edge case?

File: tests/server/test_compilerservice.py (hunk 3 above)
Comment (sanderr) [reply tests/server/test_compilerservice.py:637]: First I used the `croniter` library but it places the seconds at the end instead of the front so I switched libraries. Forgot to update this cron string.

File: stubs/crontab/_crontab.pyi [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import datetime
```
[the remark this answers is not in the document]
Comment (sanderr) [reply stubs/crontab/_crontab.pyi:1]: It would work without this. This is a typing stub, which is a way to add typing information for Python modules. A lot of third party modules are collected in the [typeshed](https://github.com/python/typeshed) project, but this one is not among them so I added it here so mypy knows the methods' signatures.

File: tests/test_util.py [hunk 7]
```diff
@@ -20,38 +20,184 @@
...
+
+
+def test_cron_schedule(monkeypatch) -> None:
+    def freeze_time(time: datetime.datetime) -> None:
+        """
+        Freeze time to a given value to avoid race conditions.
+        """
+        datetime_orig = datetime.datetime
+
+        class FrozenDatetimeMeta(type(datetime_orig)):
```
[the remark this answers is not in the document]
Comment (sanderr) [reply tests/test_util.py:67]: This is pretty exotic. It's not pretty but in tests we generally allow ourselves a bit more hackish approaches. You're just asking about why the metaclass is required, right? Not the mock of the `frozentimedatemeta`?
If so, the reason is that we need to make sure we don't break `isinstance` checks. If we just replace `datetime.datetime` with `frozendatetime`, then `not isinstance(actual_datetime_object, datetime.datetime)`. The library I used at first depended on `isinstance` checks. I've switched libraries but I haven't tried to remove this yet. I'll make sure to do so.

File: src/inmanta/server/protocol.py [hunk 8]
```diff
@@ -296,7 +296,29 @@ def get_handlers(self) -> List[routing.Rule]:
 
     # utility methods for extensions developers
     def schedule(self, call: TaskMethod, interval: int = 60, initial_delay: Optional[float] = None) -> None:
-        self._sched.add_action(call, interval, initial_delay)
+        """
+        Schedule a task repeatedly with a given interval. Tasks with the same call and the same schedule are considered the
+        same. Clients that wish to be able to delete tasks should make sure to use a unique action object.
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/server/protocol.py:301]: Arnaud had a similar question [here](https://github.com/inmanta/inmanta-core/pull/4371#discussion_r905743951), to which I responded. I propose we keep the discussion in that thread.

File: src/inmanta/server/protocol.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/server/protocol.py:311]: Same answer, but good that you asked it twice, I'd forgotten that it occurred in two places.

[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/server/protocol.py:311]: This really is just a duplicate, right?

File: src/inmanta/data/__init__.py [hunk 9]
```diff
@@ -2245,6 +2257,18 @@ def to_dto(self) -> m.Environment:
...
             doc="Allow the server to compile the configuration model.",
         ),
+        AUTO_FULL_COMPILE: Setting(
+            name=AUTO_FULL_COMPILE,
+            default="",
+            typ="str",
+            validator=validate_cron,
+            doc=(
+                "Periodically run a full compile following a cron-like time-to-run specification, interpreted in UTC"
+                " (`min hour dom month dow` or macros). A compile will be requested at the scheduled time. The actual"
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/data/__init__.py:2267]: I included it because I believe it's a standard part of cron that people might expect to work. If they then read in the documentation only `min hour dom month dow` I was afraid we'd give them a false impression. But I see now there is the opposite risk as well. I'll remove the macros reference and prefix with "e.g.": "(e.g. `min hour dom month dow`)".

File: src/inmanta/util.py [hunk 10]
```diff
@@ -112,31 +115,130 @@ def handler(future):
...
+    Abstract base class for a task schedule specification. Offers methods to inspect when the task should be scheduled, relative
+    to the current time. Stateless.
+    """
+
+    @abstractmethod
+    def get_initial_delay(self) -> float:
+        """
+        Returns the number of seconds from now until this task should be scheduled for the first time.
+        """
+        ...
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/util.py:132]: The [documentation](https://docs.python.org/3/library/abc.html#abc.abstractmethod) uses ellipsis. When you inherit from `abc.ABC` (which I forgot, that part I will definitely fix), a class can't even be instantiated if it contains abstract methods. So a `NotImplementedError` seems redundant.

File: src/inmanta/util.py (hunk 1 above)
Comment (sanderr) [reply src/inmanta/util.py:151]: Probably a good idea. I'll apply this change.

File: tests/server/test_compilerservice.py (hunk 2 above)
Comment (sanderr) [reply tests/server/test_compilerservice.py:635]: It's both actually (because it applies every day. How about "... schedule in the past/far future, check ..."?

File: tests/test_util.py (hunk 4 above)
Comment (sanderr) [reply tests/test_util.py:126]: I thought I'd fully gotten into the habit of adding docstrings to all testcases. Will fix.

File: src/inmanta/server/protocol.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/server/protocol.py:311]: I see what you mean. I think the most confusing part for me was the term `action object`. The term object makes me think about an instance of a class, but here it's a function reference, isn't it?

File: src/inmanta/util.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/util.py:132]: Probably yes. But don't feel the need to change this. It's in-line with the documentation, so that's definitely good.

File: tests/server/test_compilerservice.py (hunk 2 above)
Comment (arnaudsjs) [reply tests/server/test_compilerservice.py:635]: I think I misinterpreted the comment. It makes sense now. You can ignore this comment.

File: src/inmanta/server/protocol.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/server/protocol.py:311]: Indeed it is, but those are also objects. And even more directly so when you use `partial`. But I see what you mean. I'll try to clarify this.

File: src/inmanta/server/services/compilerservice.py [hunk 11]
```diff
@@ -487,7 +490,7 @@ def get_dependencies(self) -> List[str]:
         return [SLICE_DATABASE]
 
     def get_depended_by(self) -> List[str]:
-        return [SLICE_TRANSPORT]
+        return [SLICE_ENVIRONMENT, SLICE_SERVER, SLICE_TRANSPORT]
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/server/services/compilerservice.py:493]: Those slices have this one as their dependency. They're not strictly required here (unidirectional edges suffice) but I think it helps for clarity purposes.

File: tests/test_util.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (sanderr) [reply tests/test_util.py]: Turns out the new library indeed does not need it so I removed it.

**Comment by sanderr** [comment]
I think I addressed all review comments.

**Review by FloLey — APPROVED** [review]
